File: lcp.py
from typing import Literal


def lcp(a) -> None:
    """
    Finds the longest common prefix (LCP) in a list of strings.

    If no common prefix is found, returns an empty string.

    :param a: A list of strings to find the LCP in
    :type a: list[str]
    :return: The longest common prefix in the list of strings
    :rtype: str
    """
    prefixes: dict = {}

    for word in a:
        prefix: Literal[""] = ""
        for char in word:
            prefix += char
            prefixes.setdefault(prefix, 0)
            prefixes[prefix] += 1

    prefixesMax = max(prefixes.values())

    if prefixesMax < len(a):
        return ""

    result = max((k for k, v in prefixes.items() if v == prefixesMax), key=len)

    return result


lcp(["abc", "abcc", "abcd", "bcd"])

File: lcp.rb
def lcp(strings)
  prefixes = Hash.new { |k, v| k[v] = 0 }

  strings.each do |string|
    prefix_size = 0

    while prefix_size < string.size
      prefix = string[..prefix_size]
      prefixes[prefix] += 1
      prefix_size += 1
    end
  end
end

p lcp(["somehow", "somewhat", "something", "some"])
